File: src/main/java/ci/ufrmi/nutrici/metier/PanierService.java
package ci.ufrmi.nutrici.metier;

import ci.ufrmi.nutrici.donnees.Produit;
import ci.ufrmi.nutrici.donnees.repo.ProduitRepository;
import ci.ufrmi.nutrici.metier.dto.ArticlePanier;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class PanierService {

    private final ProduitRepository produitRepository;

    public PanierService(ProduitRepository produitRepository) {
        this.produitRepository = produitRepository;
    }

    /**
     * Calcule le total du panier.
     * Cas limites : panier vide, quantite nulle/negative, quantite > stock.
     */
    public float calculerTotal(List<ArticlePanier> articles) {
        validerArticles(articles);

        float total = 0f;
        for (ArticlePanier article : articles) {
            Produit produit = produitRepository.findById(article.getReference())
                    .orElseThrow(() -> new MetierException(
                            "Produit introuvable : " + article.getReference()));
            total += produit.getPrixUnitaire() * article.getQuantite();
        }
        return total;
    }

    public void validerArticles(List<ArticlePanier> articles) {
        if (articles == null || articles.isEmpty()) {
            throw new MetierException("Le panier est vide.");
        }

        for (ArticlePanier article : articles) {
            if (article.getQuantite() <= 0) {
                throw new MetierException(
                        "Quantite invalide pour " + article.getReference()
                                + " (doit etre > 0).");
            }

            Produit produit = produitRepository.findById(article.getReference())
                    .orElseThrow(() -> new MetierException(
                            "Produit introuvable : " + article.getReference()));

            if (article.getQuantite() > produit.getQteStock()) {
                throw new MetierException(
                        "Stock insuffisant pour " + produit.getNom()
                                + " (demande=" + article.getQuantite()
                                + ", stock=" + produit.getQteStock() + ").");
            }
        }
    }

    public void decrementerStock(List<ArticlePanier> articles) {
        for (ArticlePanier article : articles) {
            Produit produit = produitRepository.findById(article.getReference())
                    .orElseThrow(() -> new MetierException(
                            "Produit introuvable : " + article.getReference()));
            produit.setQteStock(produit.getQteStock() - article.getQuantite());
            produitRepository.save(produit);
        }
    }
}

File: src/services/panier.service.js
// src/services/panier.service.js
// Règles métier US2 (testables sans Express)

function validerArticles(articles, produitsParRef) {
  if (!articles || articles.length === 0) {
    const err = new Error('Le panier est vide.');
    err.status = 400;
    throw err;
  }

  for (const article of articles) {
    if (!article.reference || article.quantite === undefined) {
      const err = new Error('Chaque article doit avoir reference et quantite.');
      err.status = 400;
      throw err;
    }
    if (article.quantite <= 0) {
      const err = new Error(`Quantite invalide pour ${article.reference} (doit etre > 0).`);
      err.status = 400;
      throw err;
    }

    const produit = produitsParRef[article.reference];
    if (!produit) {
      const err = new Error(`Produit introuvable : ${article.reference}`);
      err.status = 404;
      throw err;
    }

    const stock = Number(produit.qte_stock);
    if (article.quantite > stock) {
      const err = new Error(
        `Stock insuffisant pour ${produit.nom} (demande=${article.quantite}, stock=${stock}).`
      );
      err.status = 400;
      throw err;
    }
  }
}

function calculerTotal(articles, produitsParRef) {
  validerArticles(articles, produitsParRef);
  let total = 0;
  for (const article of articles) {
    const produit = produitsParRef[article.reference];
    total += Number(produit.prix_unitaire) * article.quantite;
  }
  return total;
}

module.exports = { validerArticles, calculerTotal };
